src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -681,10 +680,10 @@ internal void SetNegotiatedAlpn(IntPtr alpn, int alpnLength)
                     NetEventSource.Error(state, $"{state.TraceId} Exception occurred during handling {connectionEvent.Type} connection callback: {ex}");
                 }
 
-                if (state.ConnectTcs != null)
+                if (state.ConnectTcs != null && !state.ConnectTcs.Task.IsCompleted)
                 {
-                    state.ConnectTcs.SetException(ex);
-                    state.ConnectTcs = null;
+                    // This is opportunistic if we get exception and have ability to propagate it to caller.
+                    state.ConnectTcs.TrySetException(ex);
```

Comment:
This is the change I was hinting at when asking about the race condition.

Reply:
I think I can do better fix. Instead of checking state of the task I can simply store TaskCompletionSource in ConnectAsync to local variable and return that. There is really no reason to drag it past Connect event.